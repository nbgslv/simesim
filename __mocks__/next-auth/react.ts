const mockSession = {
  expires: new Date(Date.now() + 2 * 86400).toISOString(),
  user: {
    email: '0512345678',
    emailEmail: 'test@test.com',
    id: 'clcclcclcclcclcclc1234',
    image: null,
    name: 'foo bar',
    role: 'ADMIN',
  },
};

export const useSession = jest.fn(
  () => ({ data: mockSession, status: 'authenticated' }) // return type is [] in v3 but changed to {} in v4
);
export const signOut = jest.fn(() =>
  Promise.resolve({ url: process.env.NEXTAUTH_URL as string })
);
