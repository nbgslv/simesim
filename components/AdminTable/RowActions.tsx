import React, { ReactElement, useEffect } from 'react';
import {
  GridActionsCellItem,
  GridActionsCellItemProps,
  GridRowId,
  GridRowModel,
  GridRowModes,
  GridRowModesModel,
  GridRowParams,
  GridValidRowModel,
} from '@mui/x-data-grid';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { solid } from '@fortawesome/fontawesome-svg-core/import.macro';

type RowActionsProps<T extends GridValidRowModel> = {
  rowModesModel: GridRowModesModel;
  row: GridRowModel<T>;
  rowsActions: string[];
  setRowModesModel: React.Dispatch<React.SetStateAction<GridRowModesModel>>;
  setRows: React.Dispatch<React.SetStateAction<GridRowModel<T>[]>>;
  rows: GridRowModel<T>[];
};

const RowActions = <T extends GridValidRowModel>({
  rowModesModel,
  row,
  rowsActions,
  setRowModesModel,
  setRows,
  rows,
}: RowActionsProps<T>): ReactElement<GridActionsCellItemProps>[] => {
  const [id] = React.useState<GridRowId>(row.id);
  const [inEditMode, setInEditMode] = React.useState<boolean>(false);
  const actions: string[] | ReactElement<GridActionsCellItemProps>[] = [
    ...rowsActions,
  ];

  useEffect(() => {
    setInEditMode(
      rowModesModel[(id as unknown) as GridRowId]?.mode === GridRowModes.Edit
    );
  }, [rowModesModel, id]);

  const handleEditClick = (rowId: string | number) => () => {
    setRowModesModel({
      ...rowModesModel,
      [(rowId as unknown) as GridRowId]: { mode: GridRowModes.Edit },
    });
  };

  const handleSaveClick = (rowId: GridRowParams) => () => {
    setRowModesModel({
      ...rowModesModel,
      [(rowId as unknown) as GridRowId]: { mode: GridRowModes.View },
    });
  };

  const handleDeleteClick = (rowId: string | number) => () => {
    setRows(rows.filter((rowToDelete) => rowToDelete.id !== rowId));
  };

  const handleCancelClick = (rowId: GridRowParams<T>) => () => {
    setRowModesModel({
      ...rowModesModel,
      [(rowId as unknown) as GridRowId]: {
        mode: GridRowModes.View,
        ignoreModifications: true,
      },
    });

    const editedRow = rows.find(
      (rowEdited: GridRowModel<T>) => rowEdited.id === id
    );
    if (editedRow?.isNew) {
      setRows(rows.filter((newRow: GridRowModel<T>) => newRow.id !== id));
    }
  };

  if (inEditMode) {
    return [
      <GridActionsCellItem
        label={'שמור'}
        icon={<FontAwesomeIcon icon={solid('floppy-disk')} />}
        onClick={handleSaveClick((id as unknown) as GridRowParams)}
      />,
      <GridActionsCellItem
        label={'בטל'}
        icon={<FontAwesomeIcon icon={solid('xmark')} />}
        onClick={handleCancelClick((id as unknown) as GridRowParams)}
      />,
    ];
  }

  const saveCell = (
    <GridActionsCellItem
      label={'שמור'}
      icon={<FontAwesomeIcon icon={solid('floppy-disk')} />}
      onClick={handleSaveClick((id as unknown) as GridRowParams)}
    />
  );

  const cancelCell = (
    <GridActionsCellItem
      label={'בטל'}
      icon={<FontAwesomeIcon icon={solid('xmark')} />}
      onClick={handleCancelClick((id as unknown) as GridRowParams)}
    />
  );

  const editCell = (
    <GridActionsCellItem
      label={'ערוך'}
      icon={<FontAwesomeIcon icon={solid('pen')} />}
      onClick={handleEditClick(id)}
    />
  );

  const deleteCell = (
    <GridActionsCellItem
      label={'מחק'}
      icon={<FontAwesomeIcon icon={solid('trash')} />}
      onClick={handleDeleteClick(id)}
    />
  );

  const elements: ReactElement<GridActionsCellItemProps>[] = [];
  if (inEditMode) {
    elements.length = 0;
    elements[0] = saveCell;
    elements[1] = cancelCell;
  } else {
    const indexOfEdit = actions.indexOf('edit');
    if (indexOfEdit !== -1) {
      elements[indexOfEdit] = editCell;
    }

    const indexOfDelete = actions.indexOf('delete');
    if (indexOfDelete !== -1) {
      elements[indexOfDelete] = deleteCell;
    }
  }

  return elements;
};

export default RowActions;
