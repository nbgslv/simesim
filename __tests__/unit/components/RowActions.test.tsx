import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { GridRowModel } from '@mui/x-data-grid';
import RowActions from '../../../components/AdminTable/RowActions';

describe('RowActions', () => {
  const onDelete = jest.fn(() => Promise.resolve(true));
  const rowModesModel: GridRowModel = { '1': { mode: 'view' } };
  const row = { id: '1' };
  const rowsActions = ['edit', 'delete'];
  const setRowModesModel = jest.fn();
  const setRows = jest.fn();
  const rows = [row];

  beforeEach(() => {
    onDelete.mockClear();
    setRowModesModel.mockClear();
    setRows.mockClear();
    rowModesModel[row.id] = { mode: 'view' };
  });

  it('matches snapshot', () => {
    const { container } = render(
      // @ts-ignore
      <RowActions
        rowModesModel={rowModesModel}
        row={row}
        rowsActions={rowsActions}
        setRowModesModel={setRowModesModel}
        setRows={setRows}
        rows={rows}
        onDelete={onDelete}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders the edit and delete actions', () => {
    render(
      // @ts-ignore
      <RowActions
        rowModesModel={rowModesModel}
        row={row}
        rowsActions={rowsActions}
        setRowModesModel={setRowModesModel}
        setRows={setRows}
        rows={rows}
        onDelete={onDelete}
      />
    );

    const editButton = screen.getByLabelText('ערוך');
    const deleteButton = screen.getByLabelText('מחק');

    expect(editButton).toBeInTheDocument();
    expect(deleteButton).toBeInTheDocument();
  });

  it('renders the save and cancel actions when in edit mode', () => {
    rowModesModel[row.id] = { mode: 'edit' };

    render(
      // @ts-ignore
      <RowActions
        rowModesModel={rowModesModel}
        row={row}
        rowsActions={rowsActions}
        setRowModesModel={setRowModesModel}
        setRows={setRows}
        rows={rows}
        onDelete={onDelete}
      />
    );

    const saveButton = screen.getByLabelText('שמור');
    const cancelButton = screen.getByLabelText('בטל');

    expect(saveButton).toBeInTheDocument();
    expect(cancelButton).toBeInTheDocument();
  });

  it('calls the onDelete function when delete is clicked', async () => {
    render(
      // @ts-ignore
      <RowActions
        rowModesModel={rowModesModel}
        row={row}
        rowsActions={rowsActions}
        setRowModesModel={setRowModesModel}
        setRows={setRows}
        rows={rows}
        onDelete={onDelete}
      />
    );

    const deleteButton = screen.getByLabelText('מחק');
    await userEvent.click(deleteButton);

    expect(onDelete).toHaveBeenCalledWith(row.id);
  });

  it('calls the setRows function with the updated list of rows when delete is clicked and onDelete returns true', async () => {
    onDelete.mockResolvedValue(true);

    render(
      // @ts-ignore
      <RowActions
        rowModesModel={rowModesModel}
        row={row}
        rowsActions={rowsActions}
        setRowModesModel={setRowModesModel}
        setRows={setRows}
        rows={rows}
        onDelete={onDelete}
      />
    );

    const deleteButton = screen.getByLabelText('מחק');
    await userEvent.click(deleteButton);

    expect(setRows).toHaveBeenCalledWith([]);
  });

  it('does not call the setRows function when delete is clicked and onDelete returns false', async () => {
    onDelete.mockResolvedValue(false);

    render(
      // @ts-ignore
      <RowActions
        rowModesModel={rowModesModel}
        row={row}
        rowsActions={rowsActions}
        setRowModesModel={setRowModesModel}
        setRows={setRows}
        rows={rows}
        onDelete={onDelete}
      />
    );

    const deleteButton = screen.getByLabelText('מחק');
    await userEvent.click(deleteButton);

    expect(setRows).not.toHaveBeenCalled();
  });

  it('should call setRowModesModel when edit button is clicked', () => {
    render(
      // @ts-ignore
      <RowActions
        rowModesModel={rowModesModel}
        row={row}
        rowsActions={rowsActions}
        setRowModesModel={setRowModesModel}
        setRows={setRows}
        rows={rows}
        onDelete={onDelete}
      />
    );
    userEvent.click(screen.getByLabelText('ערוך'));
    waitFor(() => expect(setRowModesModel).toHaveBeenCalled());
  });

  it('should call setRowModesModel when cancel button is clicked', () => {
    rowModesModel[row.id] = { mode: 'edit' };

    render(
      // @ts-ignore
      <RowActions
        rowModesModel={rowModesModel}
        row={row}
        rowsActions={rowsActions}
        setRowModesModel={setRowModesModel}
        setRows={setRows}
        rows={rows}
        onDelete={onDelete}
      />
    );
    userEvent.click(screen.getByLabelText('בטל'));
    waitFor(() => expect(setRowModesModel).toHaveBeenCalled());
  });

  it('should call setRowModesModel when save button is clicked', () => {
    rowModesModel[row.id] = { mode: 'edit' };

    render(
      // @ts-ignore
      <RowActions
        rowModesModel={rowModesModel}
        row={row}
        rowsActions={rowsActions}
        setRowModesModel={setRowModesModel}
        setRows={setRows}
        rows={rows}
        onDelete={onDelete}
      />
    );
    userEvent.click(screen.getByLabelText('שמור'));
    waitFor(() => expect(setRowModesModel).toHaveBeenCalled());
  });
});
