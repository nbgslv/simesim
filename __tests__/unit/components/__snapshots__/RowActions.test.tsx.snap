// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RowActions matches snapshot 1`] = `
<div>
  <button
    aria-label="ערוך"
    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
    role="menuitem"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
  <button
    aria-label="מחק"
    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
    role="menuitem"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</div>
`;
